mllib/src/main/scala/org/apache/spark/ml/feature/RobustScaler.scala

```diff
@@ -147,49 +146,73 @@ class RobustScaler (override val uid: String)
 
   override def fit(dataset: Dataset[_]): RobustScalerModel = {
     transformSchema(dataset.schema, logging = true)
-    val localRelativeError = $(relativeError)
 
-    val summaries = dataset.select($(inputCol)).rdd.map {
-      case Row(vec: Vector) => vec
-    }.mapPartitions { iter =>
-      var agg: Array[QuantileSummaries] = null
-      while (iter.hasNext) {
-        val vec = iter.next()
+    val vectors = dataset.select($(inputCol)).rdd.map { case Row(vec: Vector) => vec }
+    val numFeatures = MetadataUtils.getNumFeatures(dataset.schema($(inputCol)))
+      .getOrElse(vectors.first().size)
+
+    val localRelativeError = $(relativeError)
+    val localUpper = $(upper)
+    val localLower = $(lower)
+
+    // each featureIndex and its QuantileSummaries
+    val summaries = if (numFeatures < 1000) {
```

Comment:
In my tests with small dataset, when numFeatures>4000, then it may cause OOM.
Since the memory overhead of a QuantileSummaries is related to the number of values it has absorbed, I set a threshold=1000 conservatively.

Reply:
I see, so you have both old and new impl here. That complexity isn't great, but it does avoid performance regression while fixing the problem.
Is it possible to add a test with a lot of features, but not much data? just to exercise the code path.

Reply:
I revisit this PR and found that the performance regression may come from two side:
1, `QuantileSummaries` are not compressed after _map_ side;
2, two-step aggregation: `aggregateByKey` & `reduceByKey` may not improve the performance as supposed;

So I come back to the inital commit, and use a trick to compress aggregators at _map_ out. Now it is faster than the previous commit but still about 10% slower than Master in the edge case.
Since this impl is concise, so I prefer it to previous ones.